crates/alloy-dyn-contract/src/instance.rs

```diff
@@ -0,0 +1,108 @@
+use alloy_dyn_abi::{DynSolValue, JsonAbiExt};
+use alloy_json_abi::JsonAbi;
+use alloy_primitives::{Address, Selector};
+use alloy_providers::provider::TempProvider;
+
+use crate::{error::Result, interface::Interface, CallBuilder};
+
+/// A handle to an Ethereum contract at a specific address.
+///
+/// A contract is an abstraction of an executable program on Ethereum. Every deployed contract has
+/// an address, which is used to connect to it so that it may receive messages (transactions).
+pub struct ContractInstance<P> {
+    address: Address,
+    provider: P,
+    interface: Interface,
+}
+
+impl<P> ContractInstance<P> {
+    /// Creates a new contract from the provided address, provider, and interface.
+    pub const fn new(address: Address, provider: P, interface: Interface) -> Self {
+        Self { address, provider, interface }
+    }
+
+    /// Returns the contract's address.
+    pub const fn address(&self) -> Address {
+        self.address
+    }
+
+    /// Returns a reference to the contract's ABI.
+    pub const fn abi(&self) -> &JsonAbi {
+        self.interface.abi()
+    }
+
+    /// Returns a reference to the contract's provider.
+    pub const fn provider_ref(&self) -> &P {
+        &self.provider
+    }
+}
+
+impl<P> ContractInstance<P>
+where
+    P: Clone,
+{
+    /// Returns a clone of the contract's provider.
+    pub fn provider(&self) -> P {
+        self.provider.clone()
+    }
+
+    /// Returns a new contract instance at `address`.
+    ///
+    /// Clones `self` internally
+    #[must_use]
+    pub fn at(&self, address: Address) -> ContractInstance<P> {
+        let mut this = self.clone();
+        this.address = address;
+        this
+    }
+}
+
+impl<P: TempProvider + Clone> ContractInstance<P> {
+    /// Returns a transaction builder for the provided function name.
+    /// If there are  multiple functions with the same name due to overloading, consider using
+    /// the [`ContractInstance::function_from_selector`] method instead, since this will use the
+    /// first match.
+    pub fn function(&self, name: &str, args: &[DynSolValue]) -> Result<CallBuilder<P>> {
```

Comment:
I would love if we had a great UX about this. Right now it'd require the user to manually do `DynSolValue::X` for each argument. The way we did it with the generics `Tokenize` in Ethers-rs was nice, but understand if it's a PITA to work with.

cc @DaniPopes

Reply:
I tried to convey that at last sync but there did not seem to be a lot of agreement